File: libs/go/parliament.go
package stonecharts

import (
	"fmt"
	"math"
	"strings"
	"unicode/utf8"
)

type hemicycleRow struct {
	radius float64
	cap    int
}

func hemicycleLayout(total int, rMin, rMax float64) ([]hemicycleRow, float64) {
	if total <= 0 {
		return nil, 2.0
	}

	nRows := int(math.Ceil((-rMin + math.Sqrt(rMin*rMin+2*(rMax-rMin)*float64(total)/math.Pi)) / (rMax - rMin)))
	if nRows < 1 {
		nRows = 1
	}
	if nRows > total {
		nRows = total
	}

	radii := make([]float64, nRows)
	if nRows == 1 {
		radii[0] = 0.5 * (rMin + rMax)
	} else {
		for k := 0; k < nRows; k++ {
			radii[k] = rMin + float64(k)*(rMax-rMin)/float64(nRows-1)
		}
	}

	rawCaps := make([]int, nRows)
	rawTotal := 0
	for k := 0; k < nRows; k++ {
		c := int(math.Floor(math.Pi * radii[k]))
		if c < 1 {
			c = 1
		}
		rawCaps[k] = c
		rawTotal += c
	}

	if rawTotal < total {
		scale := float64(total) / float64(rawTotal)
		if rawTotal == 0 {
			scale = 1
		}
		rawTotal = 0
		for k := 0; k < nRows; k++ {
			c := int(math.Ceil(float64(rawCaps[k]) * scale))
			if c < 1 {
				c = 1
			}
			rawCaps[k] = c
			rawTotal += c
		}
	}

	rows := make([]hemicycleRow, 0, nRows)
	assigned := 0
	for k := 0; k < nRows; k++ {
		cap := rawCaps[k]
		if k == nRows-1 {
			cap = total - assigned
		} else if cap > total-assigned {
			cap = total - assigned
		}
		if cap <= 0 {
			continue
		}
		rows = append(rows, hemicycleRow{radius: radii[k], cap: cap})
		assigned += cap
	}

	rowGap := (rMax - rMin) / float64(nRows)
	if nRows <= 1 {
		rowGap = rMax - rMin
	}
	dotR := rowGap * 0.35
	if dotR < 1.5 {
		dotR = 1.5
	}
	if dotR > 6.0 {
		dotR = 6.0
	}

	return rows, dotR
}

func renderParliamentSVG(spec *ChartSpec) string {
	W := spec.Width
	H := spec.Height
	theme := spec.theme
	palette := theme.Palette

	cid := spec.ID
	if cid == "" {
		cid = "sc"
	}

	a11yAttr := ""
	a11yDesc := ""
	if spec.a11yOn() {
		sum := esc(a11ySummary(spec, "Parliament"))
		a11yAttr = fmt.Sprintf(` role="img" aria-label="%s"`, sum)
		a11yDesc = "<desc>" + sum + "</desc>"
	}

	mTop := 20.0
	if spec.Title != "" {
		mTop += 26
	}
	if spec.Subtitle != "" {
		mTop += 18
	}
	mLeft := 22.0
	mRight := 22.0
	mBottom := 28.0
	if spec.Legend == nil || *spec.Legend {
		mBottom += 18
	}
	if spec.Layout != nil && spec.Layout.Margin != nil {
		mm := spec.Layout.Margin
		if mm.Top != nil {
			mTop = *mm.Top
		}
		if mm.Left != nil {
			mLeft = *mm.Left
		}
		if mm.Right != nil {
			mRight = *mm.Right
		}
		if mm.Bottom != nil {
			mBottom = *mm.Bottom
		}
	}

	plotW := float64(W) - mLeft - mRight
	plotH := float64(H) - mTop - mBottom

	cx := mLeft + plotW/2
	cy := mTop + plotH*0.92

	rMax := math.Min(plotW/2, plotH*0.85)
	rMin := rMax * 0.35

	bg := "#ffffff"
	if theme.Background != "" {
		bg = theme.Background
	}
	fgTitle := theme.TitleColor
	fgSub := theme.SubtitleColor
	fgLegend := theme.LegendTextColor

	var data []float64
	var sName string
	if len(spec.Series) > 0 {
		data = spec.Series[0].Data
		sName = spec.Series[0].Name
	}
	_ = sName

	cats := spec.XAxis.Categories
	nCats := len(data)
	if len(cats) > nCats {
		nCats = len(cats)
	}
	for len(data) < nCats {
		data = append(data, 0)
	}
	for len(cats) < len(data) {
		cats = append(cats, fmt.Sprintf("%d", len(cats)))
	}

	intData := make([]int, nCats)
	total := 0
	for j := 0; j < nCats; j++ {
		v := int(math.Round(data[j]))
		if v < 0 {
			v = 0
		}
		intData[j] = v
		total += v
	}

	rows, dotR := hemicycleLayout(total, rMin, rMax)

	colors := make([]string, nCats)
	for j := 0; j < nCats; j++ {
		colors[j] = palette[j%len(palette)]
	}

	catAssignments := make([]int, 0, total)
	for j := 0; j < nCats; j++ {
		for c := 0; c < intData[j]; c++ {
			catAssignments = append(catAssignments, j)
		}
	}

	var p strings.Builder

	resp := ""
	if spec.Responsive {
		resp = fmt.Sprintf(` style="max-width:%dpx"`, W)
	}
	fmt.Fprintf(&p, `<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 %d %d" class="sc-chart" font-family="Segoe UI,Helvetica,Arial,sans-serif"%s%s>`, W, H, a11yAttr, resp)
	if a11yDesc != "" {
		p.WriteString(a11yDesc)
	}

	fmt.Fprintf(&p, `<defs><clipPath id="%s-clip"><rect x="0" y="0" width="%d" height="%d"/></clipPath></defs>`, esc(cid), W, H)

	fmt.Fprintf(&p, `<rect class="sc-bg" x="0" y="0" width="%d" height="%d" fill="%s"/>`, W, H, bg)

	ty := 18
	if spec.Title != "" {
		fmt.Fprintf(&p, `<text class="sc-title" x="%s" y="%d" text-anchor="middle" font-size="16" font-weight="600" fill="%s">%s</text>`,
			f1(float64(W)/2), ty, fgTitle, esc(spec.Title))
		ty += 20
	}
	if spec.Subtitle != "" {
		fmt.Fprintf(&p, `<text class="sc-subtitle" x="%s" y="%d" text-anchor="middle" font-size="12" fill="%s">%s</text>`,
			f1(float64(W)/2), ty, fgSub, esc(spec.Subtitle))
	}

	dotIdx := 0
	for _, row := range rows {
		for i := 0; i < row.cap; i++ {
			if dotIdx >= total {
				break
			}
			angle := math.Pi - (float64(i)+0.5)*math.Pi/float64(row.cap)
			dx := cx + row.radius*math.Cos(angle)
			dy := cy - row.radius*math.Sin(angle)
			catJ := catAssignments[dotIdx]
			fill := colors[catJ]
			fmt.Fprintf(&p, `<circle class="sc-parliament-dot sc-point" data-category="%d" data-index="%d" data-color="%s" cx="%.1f" cy="%.1f" r="%s" fill="%s"/>`,
				catJ, dotIdx, esc(fill), dx, dy, fmtNum(dotR), esc(fill))
			dotIdx++
		}
	}

	showLegend := spec.Legend == nil || *spec.Legend
	if showLegend {
		legendY := float64(H) - mBottom + 14
		totalLW := nCats * 80
		lx := (float64(W) - float64(totalLW)) / 2
		fmt.Fprintf(&p, `<g class="sc-legend" transform="translate(%.1f,%.1f)">`, lx, legendY)
		for j := 0; j < nCats; j++ {
			catLabel := cats[j]
			tLen := utf8.RuneCountInString(catLabel)
			_ = tLen
			fmt.Fprintf(&p, `<g class="sc-legend-item" data-series="%d" transform="translate(%d,0)">`, j, j*80)
			fmt.Fprintf(&p, `<rect x="0" y="0" width="10" height="10" rx="2" fill="%s"/>`, esc(colors[j]))
			fmt.Fprintf(&p, `<text x="14" y="9" font-size="11" fill="%s">%s</text>`, fgLegend, esc(catLabel))
			p.WriteString("</g>")
		}
		p.WriteString("</g>")
	}

	p.WriteString("</svg>")
	return p.String()
}

File: libs/python/stonecharts/charts/parliament.py
"""Parliament / hemicycle chart renderer: ChartSpec -> SVG string.

Unit dots arranged in concentric semicircular arcs — one dot per item,
colored by category. Single-series only; total dots = sum of values.
Non-Cartesian (Family B polar sibling) — own SVG shell, no axes.
See charts/parliament/design.md for the full geometry contract.
"""

from __future__ import annotations

import math

from ..spec import ChartSpec
from ..util import esc, fmt_num
from ._cartesian import a11y_summary


def _hemicycle_layout(total: int, r_min: float, r_max: float):
    """Compute row radii and per-row capacities for a hemicycle of *total* dots.

    Returns (rows, dot_r) where rows is a list of (radius, capacity) tuples
    and dot_r is the rendered dot radius.
    """
    if total <= 0:
        return [], 2.0

    n_rows = max(1, math.ceil((-r_min + math.sqrt(r_min**2 + 2 * (r_max - r_min) * total / math.pi)) / (r_max - r_min)))
    n_rows = max(1, min(n_rows, total))

    if n_rows == 1:
        radii = [0.5 * (r_min + r_max)]
    else:
        radii = [r_min + k * (r_max - r_min) / (n_rows - 1) for k in range(n_rows)]

    raw_caps = [max(1, math.floor(math.pi * r)) for r in radii]
    raw_total = sum(raw_caps)

    if raw_total < total:
        scale = total / raw_total if raw_total > 0 else 1
        raw_caps = [max(1, math.ceil(c * scale)) for c in raw_caps]

    rows: list[tuple[float, int]] = []
    assigned = 0
    for k in range(n_rows):
        cap = total - assigned if k == n_rows - 1 else min(raw_caps[k], total - assigned)
        if cap <= 0:
            continue
        rows.append((radii[k], cap))
        assigned += cap

    row_gap = (r_max - r_min) / n_rows if n_rows > 1 else r_max - r_min
    dot_r = max(1.5, min(6.0, row_gap * 0.35))

    return rows, dot_r


def render_svg(spec: ChartSpec) -> str:
    W, H = spec.width, spec.height
    theme = spec.theme
    palette = theme.palette
    _cid = esc(spec.id)

    a11y_attr = ""
    a11y_desc = ""
    if spec.a11y:
        _sum = esc(a11y_summary(spec, "Parliament"))
        a11y_attr = f' role="img" aria-label="{_sum}"'
        a11y_desc = f"<desc>{_sum}</desc>"

    m_top: float = 20
    if spec.title:
        m_top += 26
    if spec.subtitle:
        m_top += 18
    m_left: float = 22
    m_right: float = 22
    m_bottom: float = 28 + (18 if spec.legend else 0)
    if spec.layout and spec.layout.margin:
        m = spec.layout.margin
        if m.top is not None:
            m_top = m.top
        if m.left is not None:
            m_left = m.left
        if m.right is not None:
            m_right = m.right
        if m.bottom is not None:
            m_bottom = m.bottom

    plot_x, plot_y = m_left, m_top
    plot_w = W - m_left - m_right
    plot_h = H - m_top - m_bottom

    cx = plot_x + plot_w / 2
    cy = plot_y + plot_h * 0.92

    r_max = min(plot_w / 2, plot_h * 0.85)
    r_min = r_max * 0.35

    bg = theme.background or "#ffffff"
    fg_title = theme.title_color
    fg_sub = theme.subtitle_color
    fg_legend = theme.legend_text_color

    s0 = spec.series[0] if spec.series else None
    data = list(s0.data) if s0 else []
    cats = spec.x_axis.categories or []
    n_cats = max(len(data), len(cats))
    while len(data) < n_cats:
        data.append(0)
    while len(cats) < len(data):
        cats.append(str(len(cats)))

    int_data = [max(0, round(v)) for v in data]
    total = sum(int_data)

    rows, dot_r = _hemicycle_layout(total, r_min, r_max)

    colors = [palette[j % len(palette)] for j in range(n_cats)]

    cat_assignments: list[int] = []
    for j, count in enumerate(int_data):
        cat_assignments.extend([j] * count)

    p: list[str] = []

    vb = f"0 0 {W} {H}"
    resp = f' style="max-width:{W}px"' if spec.responsive else ""
    p.append(
        f'<svg xmlns="http://www.w3.org/2000/svg" viewBox="{vb}" '
        f'class="sc-chart" font-family="Segoe UI,Helvetica,Arial,sans-serif"'
        f"{a11y_attr}{resp}>"
    )
    if a11y_desc:
        p.append(a11y_desc)

    p.append(f'<defs><clipPath id="{_cid}-clip"><rect x="0" y="0" width="{W}" height="{H}"/></clipPath></defs>')

    p.append(f'<rect class="sc-bg" x="0" y="0" width="{W}" height="{H}" fill="{bg}"/>')

    ty = 18
    if spec.title:
        p.append(
            f'<text class="sc-title" x="{W / 2:.1f}" y="{ty}" '
            f'text-anchor="middle" font-size="16" font-weight="600" '
            f'fill="{fg_title}">{esc(spec.title)}</text>'
        )
        ty += 20
    if spec.subtitle:
        p.append(
            f'<text class="sc-subtitle" x="{W / 2:.1f}" y="{ty}" '
            f'text-anchor="middle" font-size="12" fill="{fg_sub}">'
            f"{esc(spec.subtitle)}</text>"
        )

    dot_idx = 0
    for radius, cap in rows:
        for i in range(cap):
            if dot_idx >= total:
                break
            angle = math.pi - (i + 0.5) * math.pi / cap
            dx = cx + radius * math.cos(angle)
            dy = cy - radius * math.sin(angle)
            cat_j = cat_assignments[dot_idx]
            fill = colors[cat_j]
            p.append(
                f'<circle class="sc-parliament-dot sc-point" '
                f'data-category="{cat_j}" data-index="{dot_idx}" '
                f'data-color="{esc(fill)}" '
                f'cx="{dx:.1f}" cy="{dy:.1f}" r="{fmt_num(dot_r)}" '
                f'fill="{esc(fill)}"/>'
            )
            dot_idx += 1

    if spec.legend:
        legend_y = H - m_bottom + 14
        legend_items: list[str] = []
        for j in range(n_cats):
            legend_items.append(
                f'<rect x="0" y="0" width="10" height="10" rx="2" fill="{esc(colors[j])}"/>'
                f'<text x="14" y="9" font-size="11" fill="{fg_legend}">'
                f"{esc(cats[j])}</text>"
            )
        total_w = n_cats * 80
        lx = (W - total_w) / 2
        p.append(f'<g class="sc-legend" transform="translate({lx:.1f},{legend_y:.1f})">')
        for j, item in enumerate(legend_items):
            p.append(
                f'<g class="sc-legend-item" data-series="{j}" '
                f'transform="translate({j * 80},0)">{item}</g>'
            )
        p.append("</g>")

    p.append("</svg>")
    return "".join(p)
